### numpy
numpy是Python中科学计算的基础包。它是一个Python库，提供多维数组对象、各种派生对象（例如掩码数组和矩阵）以及用于对数组进行快速操作的各种方法，包括数学、逻辑、形状操作、排序、选择、I/O 、离散傅里叶变换、基本线性代数、基本统计运算、随机模拟等等。  
numpy的部分功能如下：
- ndarray，一个具有矢量算术运算和复杂广播能力的快速且节省空间的多维数组。
- 用于对整组数据进行快速运算的标准数学函数（无需编写循环）。
- 用于读写磁盘数据的工具以及用于操作内存映射文件的工具。
- 线性代数、随机数生成以及傅里叶变换功能。
- 用于集成由C、C++、Fortran等语言编写的代码的API。

In [1]:
import numpy as np

## 一. ndarray
### 1. 核心属性
shape: 形状  
ndim：维度数量  
size：总元素个数  
dtype：元素类型

In [2]:
arr = np.array(5)
print(arr)
print("数组的形状："+str(arr.shape),"数组的维度："+str(arr.ndim))

5
数组的形状：() 数组的维度：0


In [3]:
arr = np.array([1,2,3,4])
print(arr)
print("数组的形状："+str(arr.shape),"数组的维度："+str(arr.ndim))

[1 2 3 4]
数组的形状：(4,) 数组的维度：1


In [4]:
arr = np.array([[1,2,3],[4,5,6]])
print(arr)
print("数组的形状："+str(arr.shape),"数组的维度："+str(arr.ndim))

[[1 2 3]
 [4 5 6]]
数组的形状：(2, 3) 数组的维度：2


同质性，不同的数据类型会被强制转换成相同的数据类型

In [5]:
arr = np.array([1,'hello'])
print(arr)
print(arr.dtype)

['1' 'hello']
<U11


### 2. 进阶属性  
T：转置  
flags：内存存储方式，是否连续存储  
itemsize: 单个元素占用的内存字节数  
nbytes：数组总内存占用量（size*itemsize）

In [6]:
arr = np.array([[1,2,3],[4,5,6],[7,8,9],[10,11,12]])
print('原数组：',arr,'维度：',arr.ndim,'形状：',arr.shape)
print()
print('转置数组：',arr.T,'维度：',arr.T.ndim,'形状：',arr.T.shape)

原数组： [[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]] 维度： 2 形状： (4, 3)

转置数组： [[ 1  4  7 10]
 [ 2  5  8 11]
 [ 3  6  9 12]] 维度： 2 形状： (3, 4)


### 3. ndaary的创建  
1. 基础构造：手动构建小规模或复制已有数据  
2. 预定义形状填充：快速初始化固定形状  
3. 基于数值范围生成：生成数值序列  
4. 特殊矩阵生成：数学运算用，如单位阵  
5. 随机数组生成：模拟实验数据，初始化神经网络权重  
6. 高级构造方法：处理非结构化数据或通过函数生成

In [7]:
# 从列表创建一维数组
arr1 = np.array([1, 2, 3, 4, 5])
print(arr1)
# 从嵌套列表创建二维数组
arr2 = np.array([[1, 2, 3], [4, 5, 6]])
print(arr2)
# 指定数据类型
arr3 = np.array([1, 2, 3], dtype=np.float32)
print(arr3) 

[1 2 3 4 5]
[[1 2 3]
 [4 5 6]]
[1. 2. 3.]


In [8]:
# 通过复制现有数组创建新数组
arr4 = np.array(arr1, copy=True)
print(arr4)
arr1[0] = 10
print("修改后的arr1:", arr1)
print("arr4保持不变:", arr4)

[1 2 3 4 5]
修改后的arr1: [10  2  3  4  5]
arr4保持不变: [1 2 3 4 5]


In [9]:
# 预定义形状：全0，全1，未初始化，固定值
arr = np.zeros((2,3))
print("全0数组：\n",arr)
arr = np.ones((2,3))
print("全1数组：\n",arr)
arr = np.empty((2,3))
print("未初始化数组：\n",arr)
arr = np.full((2,3),7)
print("固定值数组：\n",arr)

全0数组：
 [[0. 0. 0.]
 [0. 0. 0.]]
全1数组：
 [[1. 1. 1.]
 [1. 1. 1.]]
未初始化数组：
 [[1. 1. 1.]
 [1. 1. 1.]]
固定值数组：
 [[7 7 7]
 [7 7 7]]


In [10]:
arr = np.ones((10,))
print(arr)
print(arr.shape)  # 输出: (10,)
arr = np.ones((1,10))
print(arr)
print(arr.shape)  # 输出: (1, 10)

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
(10,)
[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]
(1, 10)


In [11]:
arr1 = np.array([1,2,3,4])
arr2 = arr1.view()
arr1[0] = 10
print("修改后的arr1:", arr1)    
print("arr2也被修改:", arr2)
arr3 = arr1.copy()
arr1[1] = 20
print("修改后的arr1:", arr1)
print("arr3保持不变:", arr3)
arr4 = np.empty_like(arr1)
print("arr4:", arr4)
arr5 = np.full_like(arr1,2026)
print("arr5:", arr5)

修改后的arr1: [10  2  3  4]
arr2也被修改: [10  2  3  4]
修改后的arr1: [10 20  3  4]
arr3保持不变: [10  2  3  4]
arr4: [1768320623 1634891111 1852795252 1701736260]
arr5: [2026 2026 2026 2026]


未初始化的数组 (np.empty)
返回给定形状和类型的未初始化的新数组。
创建未初始化的数组（值随机，取决于内存状态），适用于对性能要求极高的场景  

不推荐直接使用：需手动填充数据，否则可能引入不可预测的错误。
需要注意的是，np.empty 并不保证数组元素被初始化为 0，它只是分配内存空间，数组中的元素值是未初始化的，可能是内存中的任意值。

#### 基于数值范围生成
1. **np.arange(start, stop, step)**  
- 生成从 start 到 stop（不包含）的等差数列，步长为 step。
2. **np.linspace(start, stop, num)**  
- 生成从 start 到 stop 的 num 个等间距数值。
3. **np.logspace(start, stop, num, base)**  
- 生成从 base^start 到 base^stop 的 num 个等比数值。

In [12]:
# 等差数列
arr = np.arange(0, 10, 2)
print("等差数列：", arr)
arr = np.linspace(0, 1, num=5)
print("等间隔数列：", arr)
# 等比数列
arr = np.logspace(1, 3, num=3, base=2)
print("等比数列：", arr)

等差数列： [0 2 4 6 8]
等间隔数列： [0.   0.25 0.5  0.75 1.  ]
等比数列： [2. 4. 8.]


#### 特殊矩阵的创建  
1. 单位矩阵  
identity_matrix = np.eye(3)
2. 对角矩阵  
diagonal_matrix = np.diag([1, 2, 3])


In [15]:
identity_matrix = np.eye(3,dtype=int)
print("单位矩阵：\n", identity_matrix)
diagonal_matrix = np.diag([1, 2, 3])
print("对角矩阵：\n", diagonal_matrix)

单位矩阵：
 [[1 0 0]
 [0 1 0]
 [0 0 1]]
对角矩阵：
 [[1 0 0]
 [0 2 0]
 [0 0 3]]


#### 随机数组的生成
1. 均匀分布随机数  
uniform_random = np.random.rand(2, 3)
2. 正态分布随机数  
normal_random = np.random.randn(2, 3)
3. 整数随机数  
integer_random = np.random.randint(0, 10, size=(2, 3))
4. 指定范围随机数  
random.uniform()：返回给定形状的数组，用从低位(包含)到高位(不包含)上均匀分布的随机浮点数填充。

In [21]:
arr = np.random.rand(2,3)
print(arr)
arr1 = np.random.uniform(10,20,(2,3))
print(arr1)
arr2 = np.random.randint(0,10,(2,3))
print(arr2)
arr3 = np.random.normal(0,1,(2,3)) # 均值为0，标准差为1
print(arr3)
arr4 = np.random.randn(2,3) # 标准正态分布
print(arr4)

[[0.24191053 0.20074731 0.62778399]
 [0.7267837  0.86805653 0.82516042]]
[[14.81140012 17.04084433 18.15873637]
 [11.79176998 17.34107385 19.34546972]]
[[6 2 2]
 [3 2 5]]
[[-0.29480336 -0.10834968 -1.19333143]
 [-1.2434938  -1.04883523  0.25630119]]
[[-1.96116295 -1.4341211  -0.62859955]
 [ 1.73370408 -1.34922253  1.874983  ]]


设置随机种子（np.random.seed()）  
np.random.seed(42)  
np.random.seed 是 NumPy 中用于设置**随机数生成器种子**的函数，目的是确保程序的随机操作可以重复生成相同的结果（即保证随机结果的确定性）。这在实验复现、调试和教学场景中非常有用。


In [ ]:
np.random.seed(42)
random_numbers = np.random.rand(5)
print("使用种子42生成的随机数：", random_numbers)

使用种子42生成的随机数： [0.37454012 0.95071431 0.73199394 0.59865848 0.15601864]


In [ ]:
np.random.seed(200)
random_numbers = np.random.rand(5)
print("使用种子200生成的随机数：", random_numbers)

使用种子200生成的随机数： [0.94763226 0.22654742 0.59442014 0.42830868 0.76414069]


***种子对后续所有基于 NumPy 的随机函数生效（如 np.random.rand(), np.random.normal(), np.random.shuffle() 等）。***

### 4. ndarray的数据类型
| dtype          | 说明                             |类型代码|
|----------------|----------------------------------|-------|
| np.int32       | 32位整数                         |i4      |
| np.float64     | 64位浮点数                       |f8      |
| np.complex128  | 128位复数                        |c16     |
| np.bool_       | 布尔类型                         |        |
| np.unit32     | 32位无符号整数                    |u4      |
| np.complex64   | 64位复数                         |c8    |

In [27]:
# 显式指定数据类型
arr_int = np.array([1, 2, 3], dtype=np.int32)
# 类型转换
arr_float = arr_int.astype(np.float64)
print("整数数组：", arr_int, "数据类型：", arr_int.dtype)
print("浮点数组：", arr_float, "数据类型：", arr_float.dtype)

整数数组： [1 2 3] 数据类型： int32
浮点数组： [1. 2. 3.] 数据类型： float64


### 5. 索引与切片
| 索引/切片类型 | 描述/用法 |
|---------------|-----------|
| 基本索引     | 使用整数索引访问单个元素，如arr[0]访问第一行 |
| 切片         | 使用切片语法访问子数组，如arr[0:2, 1:3]访问前两行的第2和第3列 |
|使用slice函数  | 使用slice对象进行切片，如arr[slice(0,2), slice(1,3)] |
|布尔索引     | 使用布尔数组进行索引，如arr[arr > 5]访问大于5的元素 |

1.切片语法规则：start:stop:step，左闭右开区间（包含 start，不包含 stop）。
2.布尔索引限制：布尔数组的维度需与原数组匹配。  
3.内存共享特性：切片返回的是原数组的视图（共享内存），修改切片会影响原数组。需用 copy() 创建独立副本。

In [34]:
arr = np.random.randint(1,100,50)
arr1 = arr.reshape(5,10)
print(arr1)
print(arr1[1,2]) # 访问第二行第三列的元素
print(arr1[:,1]) # 访问第二列的所有元素
print(arr1[::2,::2]) # 访问所有偶数行和偶数列的元素
print(arr[arr>50]) # 访问所有大于50的元素

[[25 12 81 11 57 61 38 56 61 70]
 [70 80 52 28  1 78 64 41 21 85]
 [71 67 50 12 35 66 49 65 45 80]
 [78 74  9 83 44 95  6 97 87 86]
 [41 89 46 40 77 56 63 40 47 19]]
52
[12 80 67 74 89]
[[25 81 57 38 61]
 [71 50 35 49 45]
 [41 46 77 63 47]]
[81 57 61 56 61 70 70 80 52 78 64 85 71 67 66 65 80 78 74 83 95 97 87 86
 89 77 56 63]


### 6. 基本运算

大小相等的数组之间的任何算术运算都会将运算应用到元素级。

In [35]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([[7, 8, 9], [10, 11, 12]])
print(arr1 + arr2)  # 元素级加法
print(arr1 - arr2)  # 元素级减法        
print(arr1 * arr2)  # 元素级乘法
print(arr1 / arr2)  # 元素级除法

[[ 8 10 12]
 [14 16 18]]
[[-6 -6 -6]
 [-6 -6 -6]]
[[ 7 16 27]
 [40 55 72]]
[[0.14285714 0.25       0.33333333]
 [0.4        0.45454545 0.5       ]]


数组与标量的算术运算会将标量值传播到各个元素，不同大小的数组之间的运算叫做广播。

In [36]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]])
print(arr1 + 100)
print(arr1 - 100)
print(arr1 * 100)
print(arr1 / 100)

[[101 102 103]
 [104 105 106]]
[[-99 -98 -97]
 [-96 -95 -94]]
[[100 200 300]
 [400 500 600]]
[[0.01 0.02 0.03]
 [0.04 0.05 0.06]]


**广播机制是 NumPy 中一个强大的特性，它允许在不同形状的数组之间进行元素级运算。广播机制的规则如下：**

1. 如果俩个数组的维度数不相同，那么小维度数组的形状将会在最左边补1。

In [ ]:
arr1 = np.array([1,2,3])  # 形状为 (3,)
arr2 = np.array([[4], [5], [6]])  # 形状为（3，1）
# 对 arr1 应用规则 1，在其形状最左边补 1，变为 (1, 3)  [[1,2,3]]
# 此时 arr1 形状 (1, 3) 和 arr2 形状 (3, 1) 满足广播条件
print(arr1 + arr2)

[[5 6 7]
 [6 7 8]
 [7 8 9]]


2. 如果俩个数组的形状在任何一个维度上都不匹配，那么数组的形状会沿着维度大小（元素个数）为1的维度开始扩展 ，（维度必须是1开始）直到所有维度都一样， 以匹配另一个数组的形状。

In [ ]:
arr3 = np.array([[1,2,3]])
arr4 = np.array([[4],[5],[6]])
print(arr3.ndim)
print(arr3.shape)
print(arr4.shape)
# arr3 沿着第0个维度扩展,将原有的一行数据复制成3行,为 (3, 3)=>[[1,2,3], [1,2,3], [1,2,3]]
# arr4 沿着第1个维度扩展, (3, 3)=>[[4,4,4], [5,5,5], [6,6,6]]
print(arr3 + arr4)

2
(1, 3)
(3, 1)
[[5 6 7]
 [6 7 8]
 [7 8 9]]


3. 如果俩个数组的形状在任何一个维度上都不匹配，并且没有任何一个维度大小等于1，那么会引发异常。

In [45]:
arr5 = np.array([1,2,3]) # 形状为（3，）
arr6 = np.array([4,5]) # 形状为（2，）
try:
    print(arr5 + arr6)
except  ValueError as e:
    print(f"错误信息：{e}")

错误信息：operands could not be broadcast together with shapes (3,) (2,) 


***矢量化运算和广播机制的关系***  
矢量化运算和广播机制是 NumPy 中两个密切相关但又有所区别的核心概念，它们共同使得 NumPy 能够高效地处理数组运算。以下是它们的关系和区别：  
| 特性  | 矢量化运算 |广播机制 |
|-------|----------|---------|
|目标   |避免显式循环，提升计算效率|解决不同形状数组的运算问题|
|依赖关系|广播机制是矢量化运算的补充|广播机制使矢量化运算更灵活|
|形状要求|通常要求数组形状相同|允许形状不同，但需满足广播规则|
|底层实现|基于优化过的 C/Fortran 代码|自动扩展数组，不实际复制数据|


#### 矩阵乘法：  
通过*运算符和np.multiply()对两个数组相乘进行的是对位乘法而非矩阵乘法运算。

In [47]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]])
arr2 = np.array([[7, 8, 9], [10, 11, 12]])
print(arr1 * arr2)
print(np.multiply(arr1,arr2))

[[ 7 16 27]
 [40 55 72]]
[[ 7 16 27]
 [40 55 72]]


**使用np.dot()、ndarray.dot()、@可以进行矩阵乘法运算。**

In [48]:
arr1 = np.array([[1, 2, 3], [4, 5, 6]]) # (2,3)
arr2 = np.array([[6,5],[4,3],[2,1]]) # (3,2)
print(np.dot(arr1, arr2))
print(arr1 @ arr2)

[[20 14]
 [56 41]]
[[20 14]
 [56 41]]


### 7. 常用函数

|分类|	函数|	功能说明|
|----|------|----------|
|基本数学	|np.sqrt(x)|	计算平方根|
|	|np.exp(x)|	计算指数（e^x）|
 | |np.log(x)|	计算自然对数（ln(x)）|
|	|np.sin(x)	|计算正弦值|
|	|np.abs(x)	|计算绝对值|
|	|np.power(a, b)|	计算 a 的 b 次幂|
|	|np.round(x, n)|	四舍五入（保留 n 位小数）|
|统计|	np.sum(x)	|求和|
|	|np.mean(x)	|计算均值|
|	|np.median(x)	|计算中位数|
|	|np.std(x)	|计算标准差|
|	|np.var(x)	|计算方差|
|	|np.percentile(x, q)	|计算分位数（q: 0~100）|
|比较 |np.min(x)/np.max(x)  	|查找最小值/最大值|
 |   |np.greater(a, b)	|元素级比较 a > b|
|	|np.less(a, b)	|元素级比较 a < b|
|	|np.equal(a, b)	|元素级比较 a == b|
|	|np.logical_and(a, b)	|逻辑与（逐元素）|
|	|np.where(condition, x, y)	|根据条件选择元素|
|排序|	np.sort(x)	|返回排序后的副本|
|	|x.sort()	|原地排序（修改原数组）|
|	|np.argsort(x)	|返回排序后的索引|
|	|np.lexsort(keys)	|多键排序（按最后一列优先）|
|去重|	np.unique(x)	|返回唯一值并排序|
|	|np.in1d(a, b)	|检查 a 的元素是否在 b 中存在|
|其他实用|	np.concatenate((a, b))	|数组拼接|
|	|np.split(x, indices)	|分割数组|
|	|np.reshape(x, shape)	|调整数组形状|
|	|np.copy(x)	|创建数组的深拷贝|
|	|np.isnan(x)	|检测 NaN 值|

7.1 基本数学函数
| 函数 | 功能说明 | 示例代码 | 示例输出 |
| :--- | :--- | :--- | :--- |
| `np.sqrt(x)` | 计算平方根 | `np.sqrt([1, 4, 9])` | `[1. 2. 3.]` |
| `np.exp(x)` | 计算指数 ($e^x$) | `np.exp([0, 1])` | `[1. 2.71828183]` |
| `np.log(x)` | 计算自然对数 ($\ln(x)$) | `np.log([1, np.e])` | `[0. 1.]` |
| `np.sin(x)` | 计算正弦值 | `np.sin(np.pi/2)` | `1.0` |
| `np.abs(x)` | 计算绝对值 | `np.abs([-1, 2.5])` | `[1. 2.5]` |
| `np.power(a, b)` | 计算 `a` 的 `b` 次幂 | `np.power(2, [3, 4])` | `[8 16]` |
| `np.round(x, n)` | 四舍五入（保留 `n` 位小数） | `np.round(3.1415, 2)` | `3.14` |
| `np.ceil(x)` | 向上取整（不小于原数的最小整数） | `np.ceil([1.2, 3.7, -2.3])` | `[ 2. 4. -2.]` |
| `np.floor(x)` | 向下取整（不大于原数的最大整数） | `np.floor([1.2, 3.7, -2.3])` | `[ 1. 3. -3.]` |
| `np.rint(x)` | 四舍五入到最近的整数；当需要舍入的数字恰好是 5 时，会看 5 前面的数字，如果是偶数则直接舍去 5，如果是奇数则进位。 | `np.rint([1.5, 2.3, 3.7])` | `[1. 2. 4.]` |
| `np.multiply(a, b)` | 元素级乘法（等价于 `a * b`） | `np.multiply([2, 3], [4, 5])` | `[ 8 15]` |
| `np.divide(a, b)` | 元素级除法（等价于 `a / b`） | `np.divide([6, 8], [2, 4])` | `[3. 2.]` |
| `np.isnan(x)` | 检测 `NaN` 值 | `np.isnan([1, np.nan, 3])` | `[False True False]` |

7.2 统计函数
| 函数 | 功能说明 | 示例代码 | 示例输出 |
| :--- | :--- | :--- | :--- |
| `np.sum(x)` | 求和 | `np.sum([[1, 2], [3, 4]])` | `10` |
| `np.mean(x)` | 计算均值 | `np.mean([1, 2, 3, 4])` | `2.5` |
| `np.median(x)` | 计算中位数 | `np.median([5, 2, 3, 1])` | `2.5` |
| `np.std(x)` | 计算标准差 | `np.std([1, 2, 3])` | `0.81649658` |
| `np.var(x)` | 计算方差 | `np.var([1, 2, 3])` | `0.66666667` |
| `np.min(x)` | 查找最小值 | `np.min([[4, 2], [3, 1]])` | `1` |
| `np.max(x)` | 查找最大值 | `np.max([[4, 2], [3, 1]])` | `4` |
| `np.percentile(x, q)` | 计算分位数 (q: 0~100) | `np.percentile([1,2,3,4], 50)` | `2.5` |
| `np.argmax(x)` | 返回最大值的索引 | `np.argmax([1, 3, 2])` | `1` |
| `np.argmin(x)` | 返回最小值的索引 | `np.argmin([1, 3, 2])` | `0` |
| `np.cumsum(x)` | 计算累积和（逐元素累加结果） | `np.cumsum([1, 2, 3])` | `[1, 3, 6]` |
| `np.cumprod(x)` | 计算累积积（逐元素累乘结果） | `np.cumprod([1, 2, 3])` | `[1, 2, 6]` |

| 统计量名称 | NumPy方法 | 数学定义与解释 | 常见用途 | 图示理解建议 (文本) |
| :--- | :--- | :--- | :--- | :--- |
| **总和 (和)** | `np.sum()` | 所有数值相加的结果 | 总支出、总收入、总分等 | 直方图每列的"高度总和" |
| **计数** | `np.size()` | 数组中元素的总个数 (包括缺失值需额外处理) | 数据规模统计 | 数组的"总格子数" |
| **非零计数** | `np.count_nonzero()` | 数组中非零元素的数量 | 稀疏矩阵分析、激活单元统计 | 点亮非零元素的网格图 |
| **平均数** | `np.mean()` | 所有数值的总和 / 个数 | 成绩均值、产品均价 | 数轴上"重心"位置 |
| **加权平均** | `np.average(weights=)` | $\sum(\text{值} \times \text{权重}) / \sum\text{权重}$ | 加权评分、指数计算 | 不同权重的砝码平衡点 |
| **中位数** | `np.median()` | 将所有值排序后位于中间位置的值 | 房价中位数、收入中位数 | 数轴上"居中"点 |
| **众数** | `scipy.stats.mode()` | 出现频率最高的值 (需 SciPy 支持) | 最常见类别、最流行选项 | 柱状图中"最高柱"的位置 |
| **最大值** | `np.max()` | 数据中的最大元素 | 最高温度、最大销量 | 数轴最右端的点 |
| **最小值** | `np.min()` | 数据中的最小元素 | 最低成绩、最低价格 | 数轴最左端的点 |
| **极差** | `np.ptp()` | 最大值与最小值的差 (Peak-to-Peak) | 数据波动范围 | 数轴两端点之间的距离 |
| **方差** | `np.var()` | $\sum(\text{值}-\text{均值})^2 / n$ (`ddof=0` 为总体方差) | 波动性分析、风险评估 | 离均值越远的点 → 贡献越大 |
| **标准差** | `np.std()` | 方差的平方根 (`ddof=0` 为总体标准差) | 正态分布、偏差分析 | 均值 $\pm 1$ 个标准差 $\approx$ 约 68% 数据范围 |
| **分位数** | `np.quantile()` | 将数据分成 q 比例处的值 (如 0.25 为下四分位数) | 工资等级、考试分数线 | 25%、50%、75%点 → 形成"箱型图" |
| **百分位数** | `np.percentile()` | 分位数的百分数表示 (`quantile(0.25) = percentile(25)`) | 统计分布边界 | 标记特定百分比位置的参考线 |
| **协方差矩阵** | `np.cov()` | 变量间协方差的矩阵表示 (每对变量的协方差) | 多元统计分析、PCA | 热力图显示变量间相关性 |
| **相关系数** | `np.corrcoef()` | 标准化的协方差矩阵，范围 [-1, 1] | 特征选择、相关分析 | 颜色深浅表示相关性强弱 |
| **直方图统计** | `np.histogram()` | 将数据划分到等宽区间并计数 | 分布分析、数据分箱 | 柱状图显示各区间的频数 |
| **累积和** | `np.cumsum()` | 从第一个元素开始逐步累加的结果数组 | 累计收益、增长曲线 | 阶梯式上升的折线图 |
| **累积积** | `np.cumprod()` | 从第一个元素开始逐步累乘的结果数组 | 复利计算、连乘效应 | 指数增长的曲线 |
| **唯一值** | `np.unique()` | 返回排序后的唯一值数组 (可附加计数) | 类别统计、去重操作 | 彩色圆点图：每种颜色的数量 |
| **排序统计** | `np.sort() + 索引` | 排序后通过分位数位置获取统计量 | 非参数统计、排名分析 | 有序数轴上的位置标记 |
| **迹 (矩阵)** | `np.trace()` | 矩阵主对角线元素的和 | 线性代数运算、矩阵特征 | 矩阵对角线上元素的求和 |
| **范数** | `np.linalg.norm()` | 向量的模 (如 L2 范数为欧几里得距离) | 相似度计算、正则化 | 向量箭头的长度 |

7.3 比较函数

| 函数 | 功能说明 | 示例代码 | 示例输出 |
| :--- | :--- | :--- | :--- |
| `np.greater(a, b)` | 元素级比较 a > b | `np.greater([3, 1], [2, 2])` | `[True, False]` |
| `np.less(a, b)` | 元素级比较 a < b | `np.less([3, 1], [2, 2])` | `[False, True]` |
| `np.equal(a, b)` | 元素级比较 a == b | `np.equal([1, 2], [1, 3])` | `[True, False]` |
| `np.logical_and(a, b)` | 逻辑与（逐元素） | `np.logical_and([True, False], [True, True])` | `[True, False]` |
| `np.logical_or(a, b)` | 逻辑或（逐元素） | `np.logical_or([True, False], [False, False])` | `[True, False]` |
| `np.logical_not(a)` | 逻辑非（逐元素） | `np.logical_not([True, False])` | `[False, True]` |
| `np.any(a)` | 检查数组中是否至少有一个元素为 True | `np.any([0, 1, 0])` | `True` |
| `np.all(a)` | 检查数组中是否所有元素为 True | `np.all([1, 2, 0])` | `False` |
| `np.where(condition, x, y)` | 根据条件选择元素 | `np.where([True, False, True], 1, 0)` | `[1, 0, 1]` |

7.4 排序函数

| 函数 | 功能说明 | 示例代码 | 示例输出 |
| :--- | :--- | :--- | :--- |
| `np.sort(x)` | 返回排序后的副本 | `np.sort([3, 1, 2])` | `[1 2 3]` |
| `x.sort()` | 原地排序（修改原数组） | `arr = np.array([3,1,2]); arr.sort()` | `arr 变为 [1, 2, 3]` |
| `np.argsort(x)` | 返回排序后的索引 | `np.argsort([3, 1, 2])` | `[1, 2, 0]` |
| `np.lexsort(keys)` | 多键排序（按最后一列优先） | `np.lexsort(([2,1,3], [3,1,2]))` | `[1, 0, 2]` |

7.5 去重函数及其他实用函数

| 函数 | 功能说明 | 示例代码 | 示例输出 |
| :--- | :--- | :--- | :--- |
| `np.unique(x)` | 返回唯一值并排序 | `np.unique([2, 1, 2, 3, 1])` | `[1 2 3]` |
| `np.in1d(a, b)` | 检查 a 的元素是否在 b 中存在 | `np.in1d([1, 2, 3], [2, 4])` | `[False, True, False]` |
| `np.concatenate((a, b))` | 数组拼接 | `np.concatenate(([1, 2], [3, 4]))` | `[1, 2, 3, 4]` |
| `np.split(x, indices)` | 分割数组 | `np.split([1, 2, 3, 4], [2])` | `[array([1, 2]), array([3, 4])]` |
| `np.reshape(x, shape)` | 调整数组形状 | `np.reshape([1, 2, 3, 4], (2, 2))` | `[[1, 2], [3, 4]]` |
| `np.copy(x)` | 创建数组的深拷贝 | `b = np.copy(a)` | `b 与原数组 a 独立` |
| `np.isnan(x)` | 检测 NaN 值 | `np.isnan([1, np.nan, 3])` | `[False, True, False]` |